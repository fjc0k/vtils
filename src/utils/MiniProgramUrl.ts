import { assign } from 'lodash-uni'
import { createUrlQueryString } from './createUrlQueryString.ts'
import { parseUrlQueryString } from './parseUrlQueryString.ts'

export type MiniProgramUrlProvider = 'wechat'

export type MiniProgramUrlVersion = 'release' | 'develop' | 'trial'

export type MiniProgramUrlStringInput = `mp://${string}`

export type MiniProgramUrlJsonInput = {
  /** 提供商 */
  provider?: MiniProgramUrlProvider
  /** APPID */
  appId?: string
  /** 原始ID */
  rawId?: string
  /** 版本 */
  version?: MiniProgramUrlVersion
  /** 路径 */
  path?: string
  /** 查询参数 */
  query?: Record<string, any>
  /** 额外数据 */
  data?: any
}

export type MiniProgramUrlInput =
  | MiniProgramUrlStringInput
  | MiniProgramUrlJsonInput

/**
 * 小程序链接。
 */
export class MiniProgramUrl {
  private payload: MiniProgramUrlJsonInput = {}

  constructor(payload: MiniProgramUrlInput) {
    if (typeof payload === 'string') {
      this.update(JSON.parse(payload.substring(5)))
    } else {
      this.update(payload)
    }
  }

  private getPath() {
    return (
      (this.payload.path ? this.payload.path.substring(1) : '') +
      (this.payload.query ? `?${createUrlQueryString(this.payload.query)}` : '')
    )
  }

  update(payload: Partial<MiniProgramUrlJsonInput>): this {
    if (payload.path) {
      if (payload.path[0] !== '/') {
        payload.path = `/${payload.path}`
      }
      const [path, query] = payload.path.split('?')
      if (query) {
        payload.path = path
        payload.query = {
          ...parseUrlQueryString(query),
          ...payload.query,
        }
      }
    }
    assign(this.payload, payload)
    return this
  }

  toJson(): MiniProgramUrlJsonInput {
    return this.payload
  }

  toString(): string {
    return `mp://${JSON.stringify(this.payload)}`
  }

  toWxOpenLaunchWeappAttrs() {
    return {
      username: this.payload.rawId,
      path: this.getPath(),
      envVersion: this.payload.version,
      extraData: this.payload.data
        ? JSON.stringify(this.payload.data)
        : undefined,
    }
  }

  toWxNavigateToMiniProgramParams() {
    return {
      appId: this.payload.appId,
      path: this.getPath(),
      envVersion: this.payload.version,
      extraData: this.payload.data,
    }
  }

  static is(
    value: MiniProgramUrlStringInput,
  ): value is MiniProgramUrlStringInput {
    return value.substring(0, 5) === 'mp://'
  }
}
