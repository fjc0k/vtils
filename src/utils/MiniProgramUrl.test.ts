import { MiniProgramUrl } from './MiniProgramUrl.ts'

describe('MiniProgramUrl', () => {
  test('ok', () => {
    expect(
      new MiniProgramUrl(
        'mp://{"appId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ).toJson(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl(
        'mp://{"appId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ).toString(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl(
        'mp://{"appId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      )
        .update({
          path: '/pages/user?id=333',
        })
        .toJson(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl(
        'mp://{"appId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ).toWxOpenLaunchWeappAttrs(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl(
        'mp://{"rawId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ).toWxOpenLaunchWeappAttrs(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl({
        rawId: 'testApp',
        path: '/pages/home?from=223&se=33',
      }).toWxOpenLaunchWeappAttrs(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl({
        rawId: 'testApp',
        path: '/pages/home?from=223&se=33',
      }).toString(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl({
        rawId: 'testApp',
        path: 'pages/home?from=223&se=33',
      }).toString(),
    ).toMatchSnapshot()
    expect(
      new MiniProgramUrl({
        appId: 'testApp',
        path: '/pages/home?from=223&se=33',
      }).toWxNavigateToMiniProgramParams(),
    ).toMatchSnapshot()
    expect(
      MiniProgramUrl.is(
        'mp://{"rawId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ),
    ).toBeTrue()
    expect(
      MiniProgramUrl.is(
        // @ts-expect-error
        'https://{"rawId":"testApp","path":"/pages/index","query":{"x":"1"}}',
      ),
    ).toBeFalse()
  })
})
