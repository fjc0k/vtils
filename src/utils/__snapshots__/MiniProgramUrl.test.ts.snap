// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniProgramUrl ok 1`] = `
Object {
  "appId": "testApp",
  "path": "/pages/index",
  "query": Object {
    "x": "1",
  },
}
`;

exports[`MiniProgramUrl ok 2`] = `"mp://{\\"appId\\":\\"testApp\\",\\"path\\":\\"/pages/index\\",\\"query\\":{\\"x\\":\\"1\\"}}"`;

exports[`MiniProgramUrl ok 3`] = `
Object {
  "appId": "testApp",
  "path": "/pages/user",
  "query": Object {
    "id": "333",
  },
}
`;

exports[`MiniProgramUrl ok 4`] = `
Object {
  "envVersion": undefined,
  "extraData": undefined,
  "path": "pages/index?x=1",
  "username": undefined,
}
`;

exports[`MiniProgramUrl ok 5`] = `
Object {
  "envVersion": undefined,
  "extraData": undefined,
  "path": "pages/index?x=1",
  "username": "testApp",
}
`;

exports[`MiniProgramUrl ok 6`] = `
Object {
  "envVersion": undefined,
  "extraData": undefined,
  "path": "pages/home?from=223&se=33",
  "username": "testApp",
}
`;

exports[`MiniProgramUrl ok 7`] = `"mp://{\\"rawId\\":\\"testApp\\",\\"path\\":\\"/pages/home\\",\\"query\\":{\\"from\\":\\"223\\",\\"se\\":\\"33\\"}}"`;

exports[`MiniProgramUrl ok 8`] = `"mp://{\\"rawId\\":\\"testApp\\",\\"path\\":\\"/pages/home\\",\\"query\\":{\\"from\\":\\"223\\",\\"se\\":\\"33\\"}}"`;

exports[`MiniProgramUrl ok 9`] = `
Object {
  "appId": "testApp",
  "envVersion": undefined,
  "extraData": undefined,
  "path": "pages/home?from=223&se=33",
}
`;
